src/logging.c

```diff
@@ -99,49 +149,97 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
       }
     } else {
       fprintf(
-        stderr, "Error getting env. variable "
+        stderr, "Error getting environment variable "
         "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s\n", ret_str);
-      ret = RCUTILS_RET_INVALID_ARGUMENT;
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Error getting environment variable RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s", ret_str);
+      return RCUTILS_RET_ERROR;
+    }
+
+    // Set the default output stream for all severities to stderr so that errors
+    // are propagated immediately.
+    // The user can choose to set the output stream to stdout by setting the
+    // RCUTILS_LOGGING_USE_STDOUT environment variable to 1.
+    enum rcutils_get_env_retval retval = rcutils_get_env_var_zero_or_one(
+      "RCUTILS_LOGGING_USE_STDOUT", "use stderr",
+      "use stdout");
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_EMPTY:
+      case RCUTILS_GET_ENV_ZERO:
+        g_output_stream = stderr;
+        break;
+      case RCUTILS_GET_ENV_ONE:
+        g_output_stream = stdout;
+        break;
     }
 
-    // Check the environment variable for colorized output
-    const char * colorized_output;
-    ret_str = rcutils_get_env("RCUTILS_COLORIZED_OUTPUT", &colorized_output);
+    // Allow the user to choose how buffering on the stream works by setting
+    // RCUTILS_LOGGING_BUFFERED_STREAM.
+    // With an empty environment variable, use the default of the stream.
+    // With a value of 0, force the stream to be unbuffered.
+    // With a value of 1, force the stream to be line buffered.
+    retval =
+      rcutils_get_env_var_zero_or_one(
+      "RCUTILS_LOGGING_BUFFERED_STREAM", "not buffered",
+      "buffered");
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_EMPTY:
+        break;
+      case RCUTILS_GET_ENV_ZERO:
+        if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
+          char error_string[1024];
+          rcutils_safe_strerror(error_string, sizeof(error_string));
+          fprintf(
+            stderr, "Error setting stream buffering mode: %s\n", error_string);
+          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+            "Error setting stream buffering mode: %s", error_string);
+          return RCUTILS_RET_ERROR;
+        }
+        break;
+      case RCUTILS_GET_ENV_ONE:
+        if (setvbuf(g_output_stream, NULL, _IOLBF, 0) != 0) {
+          char error_string[1024];
+          rcutils_safe_strerror(error_string, sizeof(error_string));
+          fprintf(
+            stderr, "Error setting stream buffering mode: %s\n", error_string);
+          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+            "Error setting stream buffering mode: %s", error_string);
+          return RCUTILS_RET_ERROR;
+        }
+        break;
+    }
```

Comment:
Why no default case? I'm not sure if the compiler will warn you (certainly will not error I think) if you added a field to the enum later and didn't update the switch statements?

Reply:
Good call.  I've added a default case to all switches.